sdk/src/Core/Amazon.Runtime/Credentials/DefaultInstanceProfileAWSCredentials.cs

```diff
@@ -269,105 +278,174 @@ public override async Task<ImmutableCredentials> GetCredentialsAsync()
         #endregion
 
         #region Private members
-        private void RenewCredentials(object unused)
+        private void RenewCredentials(object _)
         {
+            var logger = Logger.GetLogger(typeof(DefaultInstanceProfileAWSCredentials));
+
+            // This would only be true for unit tests that want to disable the timer-based refresh for
+            // more predictable testing.
+            if (_refreshRate <= TimeSpan.Zero)
+                return;
+
             // By default, the refreshRate will continue to be 
             // _refreshRate, but if FetchCredentials() returns an expired credential,
             // the refresh rate will be adjusted
             var refreshRate = _refreshRate;
-
+            var lockedObtained = false;
             try
             {
-                // if FetchCredentials() call were to fail, _lastRetrievedCredentials
-                // would remain unchanged and would continue to be returned in GetCredentials()
-                _lastRetrievedCredentials = FetchCredentials();
-
-                // check for a first time failure
-                if (!_imdsRefreshFailed &&
-                    _lastRetrievedCredentials.IsExpiredWithin(TimeSpan.Zero))
+                logger.DebugFormat("[Background Timer] Waiting on lock to refresh ECS IMDS");
+                if (_credentialsSemaphore.Wait(_credentialsLockTimeout))
```

Comment:
It is likely best to check for _isDisposed (which you set early before disposing the timer) and return if set to prevent potential exceptions during disposing.

Reply:
Done